packages/backend-storage/src/storage_access_arbiter.ts

```diff
@@ -0,0 +1,145 @@
+    {
+      acceptor: ResourceAccessAcceptor;
+      accessMap: Map<
+        StorageAction,
+        { allow: Set<StoragePath>; deny: Set<StoragePath> }
+      >;
+    }
+  >();
+
+  /**
+   * Maintains pointers to the "deny" StoragePath Set for each access entry in the map above
+   * This map is used during a final pass over all the StoragePaths to deny access on any paths where explicit allow rules were not specified
+   */
+  private prefixDenyMap = new Map<StoragePath, SetDenyByDefault[]>();
+
+  /**
+   * Initialize with a policyFactory.
+   * The policyFactory is the component that actually knows how to construct IAM policies.
+   * This class only manages the defineStorage access pattern abstraction
+   */
+  constructor(private readonly policyFactory: StorageAccessPolicyFactory) {}
+
+  /**
+   * Update the translator with an access definition.
+   * This definition defines a set of actions on a single s3 prefix that should be attached to a given ResourceAccessAcceptor
+   */
+  addAccessDefinition = (
+    resourceAccessAcceptor: ResourceAccessAcceptor,
+    actions: StorageAction[],
+    s3Prefix: StoragePath
+  ) => {
+    const acceptorToken = resourceAccessAcceptor.identifier;
+
+    // if we haven't seen this token before, add it to the map
+    if (!this.acceptorAccessMap.has(acceptorToken)) {
+      this.acceptorAccessMap.set(acceptorToken, {
+        accessMap: new Map(),
+        acceptor: resourceAccessAcceptor,
+      });
+    }
+    const accessMap = this.acceptorAccessMap.get(acceptorToken)!.accessMap;
+    // add each action to the accessMap for this acceptorToken
+    actions.forEach((action) => {
+      if (!accessMap.has(action)) {
+        // if we haven't seen this action for this acceptorToken before, add it to the map
+        const allowSet = new Set<StoragePath>([s3Prefix]);
+        const denySet = new Set<StoragePath>();
+        accessMap.set(action, { allow: allowSet, deny: denySet });
+
+        // this is where we create the reverse mapping that allows us to add entries to the denySet later by looking up the prefix
+        this.setPrefixDenyMapEntry(s3Prefix, allowSet, denySet);
+      } else {
+        // otherwise add the prefix to the existing allow set
+        accessMap.get(action)?.allow.add(s3Prefix);
+      }
+    });
+  };
+
+  /**
+   * Iterates over all of the access definitions that have been added to the translator,
+   * generates a policy for each accessMap,
+   * and attaches the policy to the corresponding ResourceAccessAcceptor
+   *
+   * After this method is called, the existing access definition state is cleared.
+   * This prevents multiple calls to this method from producing duplicate policies.
+   * The class can continue to be used to build up state for a new set of policies if desired.
+   * @param ssmEnvironmentEntries Additional SSM context that is passed to each ResourceAccessAcceptor
+   */
+  attachPolicies = (ssmEnvironmentEntries: SsmEnvironmentEntry[]) => {
```

Comment:
and this is the method where we use those data structures to go through all the paths, find prefixes and update corresponding deny rules